### PR TITLE
Java: Add XXE sinks

### PR DESCRIPTION
**Pull request by haby0**
There are many XML parsers for Java, and most of them are vulnerable to XXE because their default settings enable parsing of
external entities. This query currently identifies vulnerable XML parsing from the following parsers: javax.xml.validation.Validator,
org.dom4j.DocumentHelper, org.apache.commons.digester3.Digester, java.beans.XMLDecoder, org.rundeck.api.parser.ParserHelper, org.apache.commons.digester.Digester, org.apache.tomcat.util.digester.Digester

Open source components: [vertx-web](https://github.com/vert-x3/vertx-web/blob/3.5.3/vertx-web-api-contract/src/main/java/io/vertx/ext/web/api/validation/impl/XMLTypeValidator.java#L35)

### CODE REVIEW

**Comment by smowton** [comment]
This PR is currently a big rebase and merge mess -- recommend flattening unnecessary commits to make reviewing this easier.

File: java/ql/src/experimental/Security/CWE/CWE-611/XXE.ql [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/**
+ * @name Resolving XML external entity in user-controlled data
+ * @description Parsing user-controlled XML documents and allowing expansion of external entity
+ *              references may lead to disclosure of confidential data or denial of service.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id java/xxe
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-611/XXE.ql:8]:
```suggestion
 * @id java/xxe-with-experimental-sinks
```

File: java/ql/src/experimental/Security/CWE/CWE-611/XXE.ql [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/**
+ * @name Resolving XML external entity in user-controlled data
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-611/XXE.ql:2]:
```suggestion
 * @name Resolving XML external entity in user-controlled data (experimental sinks)
```

File: java/ql/src/experimental/Security/CWE/CWE-611/XXE.ql [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * @description Parsing user-controlled XML documents and allowing expansion of external entity
+ *              references may lead to disclosure of confidential data or denial of service.
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-611/XXE.ql:4]:
```suggestion
 * @description Parsing user-controlled XML documents and allowing expansion of external entity
 *              references may lead to disclosure of confidential data or denial of service.
 *              (note this version differs from query `java/xxe` by including support for additional possibly-vulnerable XML parsers)
```

File: java/ql/src/experimental/Security/CWE/CWE-611/XXELib.qll [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    )
+  }
+
+  override Expr getSink() { result = this.getArgument(0) }
+
+  override predicate isSafe() { none() }
+}
+
+/**
+ * The classes `org.apache.commons.digester3.Digester` or `org.apache.commons.digester.Digester` or `org.apache.tomcat.util.digester.Digester`.
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-611/XXELib.qll:120]:
```suggestion
 * The classes `org.apache.commons.digester3.Digester`, `org.apache.commons.digester.Digester` or `org.apache.tomcat.util.digester.Digester`.
```

File: java/ql/src/experimental/Security/CWE/CWE-611/XXELocal.ql [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/**
+ * @name Resolving XML external entity in local source
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-611/XXELocal.ql:2]: Is there any reason to add this version? Did you need it to catch some CVE? If not, drop the local version.

File: java/ql/lib/semmle/code/java/security/XmlParsers.qll [hunk 6]
```diff
@@ -1002,7 +1002,7 @@ class SAXTransformerFactoryNewXMLFilter extends XmlParserCall {
     exists(Method m |
       this.getMethod() = m and
       m.getDeclaringType().hasQualifiedName("javax.xml.transform.sax", "SAXTransformerFactory") and
-      m.hasName("newXMLFilter")
+      m.getName() in ["newXMLFilter", "newTransformerHandler"]
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/security/XmlParsers.qll:1005]: Move this change to the experimental query

File: java/ql/lib/semmle/code/java/security/XmlParsers.qll [hunk 7]
```diff
@@ -144,7 +144,7 @@ private predicate constantStringExpr(Expr e, string val) {
 }
 
 /** An expression that always has the same string value. */
-private class ConstantStringExpr extends Expr {
+class ConstantStringExpr extends Expr {
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/security/XmlParsers.qll:147]: Copy this to your qll file rather than expose it: the heuristic is bad enough that we probably don't want to publicly expose this difficult-to-explain partial solution.

File: java/ql/src/experimental/Security/CWE/CWE-611/XXELocal.ql (hunk 5 above)
Comment (haby0) [reply java/ql/src/experimental/Security/CWE/CWE-611/XXELocal.ql:2]: Yes i am using it to catch some cve.

File: java/ql/lib/semmle/code/java/security/XmlParsers.qll [hunk 8]
```diff
@@ -991,28 +991,6 @@ class SafeTransformer extends MethodAccess {
   }
 }
 
-/*
- * SAXTransformer: https://cheatsheetseries.owasp.org/cheatsheets/XML_External_Entity_Prevention_Cheat_Sheet.html#saxtransformerfactory
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/security/XmlParsers.qll:995]: Undo this change: this file should be unchanged except exposing `SafeTransformerFactoryFlowConfig` which is acceptable, then the experimental query should add the extra sink.

File: java/ql/src/experimental/Security/CWE/CWE-611/XXELocal.ql [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ * @description Parsing user-controlled XML documents and allowing expansion of external entity
+ *              references may lead to disclosure of confidential data or denial of service.
+ * @kind path-problem
+ * @problem.severity recommendation
+ * @precision medium
+ * @id java/xxe
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-611/XXELocal.ql:8]:
```suggestion
 * @name Resolving XML external entity from a local source (experimental sinks)
 * @description Parsing user-controlled XML documents and allowing expansion of external entity
 *              references may lead to disclosure of confidential data or denial of service.
 *              (note this version differs from query `java/xxe` by including support for additional possibly-vulnerable XML parsers,
 *              and by considering local information sources dangerous (e.g. environment variables) in addition to the remote sources
 *              considered by the normal `java/xxe` query)
 * @kind path-problem
 * @problem.severity recommendation
 * @precision medium
 * @id java/xxe-local-experimental-sinks
```

**Review by smowton — APPROVED** [review]
